x-bytestring/src/X/Data/ByteString/Unsafe.hs

```diff
@@ -0,0 +1,34 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+module X.Data.ByteString.Unsafe (
+    unsafeSplits
+  , unsafeSlice
+  ) where
+
+import           Data.ByteString.Internal (ByteString(..))
+import qualified Data.Vector.Generic as Generic
+
+import           P
+
+
+data IdxOff =
+  IdxOff !Int !Int
+
+unsafeSplits :: (Generic.Vector va a, Generic.Vector vn Int) => (ByteString -> a) -> ByteString -> vn Int -> va a
+unsafeSplits f bs sizes =
+  let
+    loop (IdxOff idx off) =
+      let
+        !len =
+          Generic.unsafeIndex sizes idx
+      in
+        Just (f $! unsafeSlice off len bs, IdxOff (idx + 1) (off + len))
+  in
+    Generic.unfoldrN (Generic.length sizes) loop (IdxOff 0 0)
+{-# INLINE unsafeSplits #-}
```

Comment:
this looks like it'd be useful for confluence too: https://github.com/ambiata/confluence/blob/master/src/Confluence/Segd.hs